provider/activitypub/mastodon/client.go

```diff
@@ -1,67 +1,299 @@
 )
 
 var _ Client = (*client)(nil)
 
 type Client interface {
-	GetKafkaConsumer() *kgo.Client
+	FollowRelayServices(ctx context.Context) error
+	GetMessageChan() (<-chan *map[string]interface{}, error)
+	GetActor() (*activitypub.Actor, error)
+	SendMessage(*map[string]interface{})
+	FetchAnnouncedObject(ctx context.Context, objectURL string) (*activitypub.Object, error)
+	// GetKafkaConsumer() *kgo.Client
 }
 
 type client struct {
-	httpClient    *http.Client
-	encoder       *form.Encoder
-	attempts      uint
-	kafkaConsumer *kgo.Client
+	httpClient *http.Client
+	encoder    *form.Encoder
+	attempts   uint
+	domain     string
+	privateKey *rsa.PrivateKey
+	signer     httpsig.Signer
+	msgChan    chan *map[string]interface{}
+	actor      *activitypub.Actor
+	relayURLs  []string
 }
 
-func (c *client) GetKafkaConsumer() *kgo.Client {
-	return c.kafkaConsumer
-}
+func NewClient(endpoint string) (Client, error) {
+	privateKey, publicKeyPem, err := generateKeyPair()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create actor template
+	actor := &activitypub.Actor{
+		Context: []string{
+			"https://www.w3.org/ns/activitystreams",
+			"https://w3id.org/security/v1",
+		},
+		Type:   "Application",
+		ID:     fmt.Sprintf("%s/actor", endpoint),
+		Inbox:  fmt.Sprintf("%s/actor/inbox", endpoint),
+		Outbox: fmt.Sprintf("%s/actor/outbox", endpoint),
+		PublicKey: struct {
+			ID           string `json:"id"`
+			Owner        string `json:"owner"`
+			PublicKeyPem string `json:"publicKeyPem"`
+		}{
+			ID:           fmt.Sprintf("%s/actor#main-key", endpoint),
+			Owner:        fmt.Sprintf("%s/actor", endpoint),
+			PublicKeyPem: publicKeyPem,
+		},
+		Endpoints: struct {
+			SharedInbox string `json:"sharedInbox"`
+		}{
+			SharedInbox: fmt.Sprintf("%s/inbox", endpoint),
+		},
+	}
 
-func NewClient(endpoint string, kafkaTopic string, offset *int64) (Client, error) {
-	instance := client{
+	// Create signer
+	signer, _, err := httpsig.NewSigner(
+		[]httpsig.Algorithm{httpsig.RSA_SHA256},
+		httpsig.DigestSha256,
+		[]string{httpsig.RequestTarget, "Host", "Date", "Digest", "Content-Type"},
+		httpsig.Signature,
+		60*60,
+	)
+	if err != nil {
+		zap.L().Error("Error creating NewSigner: %v", zap.Error(err))
+	}
+
+	// Set the Relay URL list
+	relayURLs := []string{
+		"https://relay.fedi.buzz/instance/mas.to",
+		"https://relay.fedi.buzz/instance/mastodon.social",
+	}
+
+	return &client{
+		domain:     endpoint,
+		privateKey: privateKey,
+		signer:     signer,
+		msgChan:    make(chan *map[string]interface{}, MessageChannelMaxLength),
+		actor:      actor,
 		httpClient: &http.Client{
 			Timeout: DefaultTimeout,
 		},
-		encoder:  form.NewEncoder(),
-		attempts: DefaultAttempts,
+		encoder:   form.NewEncoder(),
+		attempts:  DefaultAttempts,
+		relayURLs: relayURLs,
+	}, nil
+}
+func (c *client) FetchAnnouncedObject(ctx context.Context, objectURL string) (*activitypub.Object, error) {
+	httpClient, err := httpx.NewHTTPClient()
```

Comment:
use `c.httpClient` instead of creating a new instance

Reply:
Resolved.